{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb892ad9-86fd-4fd4-95cc-dd2b0683023a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Readme\n",
    "In this computational experiment, the aim is to verify the binding of forward and reverse primers to the target sequence. The primary objective is to identify the resulting DNA sequences and their corresponding base pair (bp) bands that would be generated in a PCR experiment. This virtual experiment is a crucial step towards performing a real experiment and assessing the availability of reagents.\n",
    "\n",
    "The experiment revealed that a PCR product necessary for HIV research, particularly for subtype A1 and A6 prevalent in Kazakhstan, can consistently generated by pairing A with D, and in a few cases 3LTR with D. There are rare cases when 3LTR and 5LTR primers bind."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cb7cb3a-d4ae-4dac-b11b-57a3ba9d6dd6",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f07139c3-271c-403e-8eab-db11007d62a5",
   "metadata": {},
   "source": [
    "Information about HIV plasmid: https://www.hivreagentprogram.org/Catalog/Clones/ARP-194.aspx. This plasmid is circular.\n",
    "Some primers contain unusual letters, such as W or R. They mean W = A or T\n",
    "R = A or G. Those primers were omitted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9d120adf-2265-42ef-9c3d-ff64f4ed558f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4bde2b84-344b-49e0-adbc-1a6f7588843f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CATCAATGAGGAAGCTGCAGAATGGGATAGAGTACATCCAGTGCATGCAGGGCCTATTGCACCAGGCCAGATGAGAGAACCAAGGGGAAGTGACATAGCA...\n"
     ]
    }
   ],
   "source": [
    "# get reference sequence for HIV plasmid\n",
    "with open('Sequences/hiv_plasmid.txt') as f:\n",
    "    plasmid_HIV = f.read()\n",
    "    \n",
    "len(plasmid_HIV)\n",
    "print(plasmid_HIV[:100] + '...')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4f6622a1-4f59-4730-9cce-440a89b78c06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': 'CAGGAGCAGATGATACAG',\n",
       " 'D': 'CCCTTCACCTTTCCAGAG',\n",
       " 'LSS': 'GTCCCTTAAGCGGAG',\n",
       " 'LSP': 'GTAATACGACTCACTATAGGGCCTCCGCTTAAGGGACT',\n",
       " '3LTR': 'TGTGACTCTGGTAACTAGAGATCCCTC',\n",
       " '3LTRnest': 'GAGATCCCTCAGACCCTTTTAGTCAG',\n",
       " '5LTR': 'TCAGGGAAGTAGCCTTGTGTGTGGT',\n",
       " '5LTRnest': 'CACTGTTGTCTTTTCTGGGAGTGAACTAGCC'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a list of primers\n",
    "primers = {'A': 'CAG GAG CAG ATG ATA CAG',\n",
    "           'D': 'CCC TTC ACC TTT CCA GAG',\n",
    "           'LSS': 'GTCCCTTAAGCGGAG',\n",
    "           'LSP': 'GTAATACGACTCACTATAGGGCCTCCGCTTAAGGGACT',\n",
    "           '3LTR': 'TGTGACTCTGGTAACTAGAGATCCCTC',\n",
    "           '3LTRnest': 'GAGATCCCTCAGACCCTTTTAGTCAG',\n",
    "           '5LTR': 'TCAGGGAAGTAGCCTTGTGTGTGGT',\n",
    "           '5LTRnest': 'CACTGTTGTCTTTTCTGGGAGTGAACTAGCC',}\n",
    "\n",
    "# clean primers\n",
    "for key in primers:\n",
    "    seq = primers[key]\n",
    "    primers[key] = \"\".join(seq.split())\n",
    "\n",
    "primers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc7bee27-3f91-4c46-ad47-59abf8d49357",
   "metadata": {},
   "source": [
    "# Function to find PCR product"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13dee699-5e57-46d5-ac87-5d40184f2cd5",
   "metadata": {},
   "source": [
    "One way to learn about a possible product is to run this experiment from https://www.bioinformatics.org/sms2/pcr_products.html. The code below achieves the same goal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8257be91-f350-439a-b494-47b1015670fa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# function that finds PCR product\n",
    "\n",
    "\n",
    "def pcr_product(forward_primer, reverse_primer, target_seq):\n",
    "    \"\"\"\n",
    "    find PCR product along with its indexes\n",
    "    treats target molecule as linear molecule\n",
    "    \"\"\"\n",
    "    \n",
    "    # target seq\n",
    "    target_seq = Seq(target_seq)\n",
    "    \n",
    "    # convert reverse primer\n",
    "    forward_primer = Seq(forward_primer)\n",
    "    reverse_primer = Seq(reverse_primer[::-1]).complement()\n",
    "    \n",
    "    # find indexes of primers\n",
    "    forward_index = target_seq.find(forward_primer)\n",
    "    reverse_index = target_seq.find(reverse_primer) + len(reverse_primer)\n",
    "    \n",
    "    # define length and a sequence of a product\n",
    "    if forward_index != -1 and reverse_index != -1 and forward_index < reverse_index:\n",
    "        product = target_seq[forward_index:reverse_index]\n",
    "        return product, (forward_index, reverse_index, reverse_index-forward_index)\n",
    "    else:\n",
    "        return [], (forward_index, reverse_index, -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5b26eed6-5f45-4b0f-8845-c7a3eb362f1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2642\n",
      "(964, 3606, 2642)\n",
      "CAGGAGCAGATGATACAGTATTAGAAGAAATGAGTTTGCCAGGAAGATGGAAACCAAAAATGATAGGGGGAATTGGAGGTTTTATCAAAGTAAGACAGTATGATCAGATACTCATAGAAATCTGTGGACATAAAGCTATAGGTACAGTATTAGTAGGACCTACACCTGTCAACATAATTGGAAGAAATCTGTTGACTCAGATTGGTTGCACTTTAAATTTTCCCATTAGCCCTATTGAGACTGTACCAGTAAAATTAAAGCCAGGAATGGATGGCCCAAAAGTTAAACAATGGCCATTGACAGAAGAAAAAATAAAAGCATTAGTAGAAATTTGTACAGAAATGGAAAAGGAAGGGAAAATTTCAAAAATTGGGCCTGAGAATCCATACAATACTCCAGTATTTGCCATAAAGAAAAAAGACAGTACTAAATGGAGAAAATTAGTAGATTTCAGAGAACTTAATAAGAGAACTCAAGACTTCTGGGAAGTTCAATTAGGAATACCACATCCCGCAGGGTTAAAAAAGAAAAAATCAGTAACAGTACTGGATGTGGGTGATGCATATTTTTCAGTTCCCTTAGATGAAGACTTCAGGAAGTATACTGCATTTACCATACCTAGTATAAACAATGAGACACCAGGGATTAGATATCAGTACAATGTGCTTCCACAGGGATGGAAAGGATCACCAGCAATATTCCAAAGTAGCATGACAAAAATCTTAGAGCCTTTTAAAAAACAAAATCCAGACATAGTTATCTATCAATACATGGATGATTTGTATGTAGGATCTGACTTAGAAATAGGGCAGCATAGAACAAAAATAGAGGAGCTGAGACAACATCTGTTGAGGTGGGGACTTACCACACCAGACAAAAAACATCAGAAAGAACCTCCATTCCTTTGGATGGGTTATGAACTCCATCCTGATAAATGGACAGTACAGCCTATAGTGCTGCCAGAAAAAGACAGCTGGACTGTCAATGACATACAGAAGTTAGTGGGGAAATTGAATTGGGCAAGTCAGATTTACCCAGGGATTAAAGTAAGGCAATTATGTAAACTCCTTAGAGGAACCAAAGCACTAACAGAAGTAATACCACTAACAGAAGAAGCAGAGCTAGAACTGGCAGAAAACAGAGAGATTCTAAAAGAACCAGTACATGGAGTGTATTATGACCCATCAAAAGACTTAATAGCAGAAATACAGAAGCAGGGGCAAGGCCAATGGACATATCAAATTTATCAAGAGCCATTTAAAAATCTGAAAACAGGAAAATATGCAAGAATGAGGGGTGCCCACACTAATGATGTAAAACAATTAACAGAGGCAGTGCAAAAAATAACCACAGAAAGCATAGTAATATGGGGAAAGACTCCTAAATTTAAACTACCCATACAAAAGGAAACATGGGAAACATGGTGGACAGAGTATTGGCAAGCCACCTGGATTCCTGAGTGGGAGTTTGTTAATACCCCTCCTTTAGTGAAATTATGGTACCAGTTAGAGAAAGAACCCATAGTAGGAGCAGAAACCTTCTATGTAGATGGGGCAGCTAACAGGGAGACTAAATTAGGAAAAGCAGGATATGTTACTAACAAAGGAAGACAAAAGGTTGTCCCCCTAACTAACACAACAAATCAGAAAACTGAGTTACAAGCAATTTATCTAGCTTTGCAGGATTCAGGATTAGAAGTAAACATAGTAACAGACTCACAATATGCATTAGGAATCATTCAAGCACAACCAGATAAAAGTGAATCAGAGTTAGTCAATCAAATAATAGAGCAGTTAATAAAAAAGGAAAAGGTCTATCTGGCATGGGTACCAGCACACAAAGGAATTGGAGGAAATGAACAAGTAGATAAATTAGTCAGTGCTGGAATCAGGAAAATACTATTTTTAGATGGAATAGATAAGGCCCAAGATGAACATGAGAAATATCACAGTAATTGGAGAGCAATGGCTAGTGATTTTAACCTGCCACCTGTAGTAGCAAAAGAAATAGTAGCCAGCTGTGATAAATGTCAGCTAAAAGGAGAAGCCATGCATGGACAAGTAGACTGTAGTCCAGGAATATGGCAACTAGATTGTACACATTTAGAAGGAAAAGTTATCCTGGTAGCAGTTCATGTAGCCAGTGGATATATAGAAGCAGAAGTTATTCCAGCAGAAACAGGGCAGGAAACAGCATATTTTCTTTTAAAATTAGCAGGAAGATGGCCAGTAAAAACAATACATACAGACAATGGCAGCAATTTCACCAGTGCTACGGTTAAGGCCGCCTGTTGGTGGGCGGGAATCAAGCAGGAATTTGGAATTCCCTACAATCCCCAAAGTCAAGGAGTAGTAGAATCTATGAATAAAGAATTAAAGAAAATTATAGGACAGGTAAGAGATCAGGCTGAACATCTTAAGACAGCAGTACAAATGGCAGTATTCATCCACAATTTTAAAAGAAAAGGGGGGATTGGGGGGTACAGTGCAGGGGAAAGAATAGTAGACATAATAGCAACAGACATACAAACTAAAGAATTACAAAAACAAATTACAAAAATTCAAAATTTTCGGGTTTATTACAGGGACAGCAGAAATCCACTTTGGAAAGGACCAGCAAAGCTCCTCTGGAAAGGTGAAGGG\n"
     ]
    }
   ],
   "source": [
    "# find PCR product\n",
    "PCR_result, indexes = pcr_product(primers['A'], primers['D'], plasmid_HIV)\n",
    "print(len(PCR_result))\n",
    "print(indexes)\n",
    "print(PCR_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cd50976d-bedd-444e-9e14-07fc29a2a31b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "(-1, 17, -1)\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "# find PCR product\n",
    "PCR_result, indexes = pcr_product(primers['D'], primers['A'], plasmid_HIV)\n",
    "print(len(PCR_result))\n",
    "print(indexes)\n",
    "print(PCR_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "999c3127-be71-45f9-889c-b2617e6876d9",
   "metadata": {},
   "source": [
    "# PCR products for HIV primers on HIV plasmid"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5315758d-e434-4882-b308-4e40916e3663",
   "metadata": {},
   "source": [
    "Serrao et al. in \"Amplification, Next-generation Sequencing, and Genomic DNA Mapping of Retroviral Integration Sites\" https://www.jove.com/t/53840/amplification-next-generation-sequencing-genomic-dna-mapping gives protocol for working with HIV virus. The protocol suggests primers for PCR amplification and linker construction. Those are LSS, 3LTR, LSP and others, see below.\n",
    "\n",
    "The primer sequences themselves do not inherently contain information about their orientation or direction. The terms \"forward\" and \"reverse\" are used to describe the directionality of the PCR amplification, which depends on the specific application and experimental setup. That is why the code below tests \"A\" and \"D\" as well as \"D\" and \"A\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "95663a66-ea52-4edd-a332-569d2d0549f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "count = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6fb8ae6f-63a6-445b-918e-bc8debe4001b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run pcr with suggested primers\n",
    "def run_pcr(suggested_primers, target_seq, target_name):\n",
    "    '''\n",
    "    Run pcr with select primers. Account for circular DNA.\n",
    "    \n",
    "    suggested_primers must be a list with length at least 2.\n",
    "    target_seq - DNA in string format\n",
    "    \n",
    "    It uses primers dict defined above\n",
    "    '''\n",
    "    global count\n",
    "    \n",
    "    target_seq_rotated = target_seq[200:] + target_seq[:200]\n",
    "    \n",
    "    if len(primers) < 2:\n",
    "        raise ValueError(\"Less than 2 primers\")\n",
    "        \n",
    "    target_seq_rotate = target_seq[200:] + target_seq[:200] # rotation to the right\n",
    "    \n",
    "    for i in suggested_primers:\n",
    "        \n",
    "        for j in suggested_primers:\n",
    "            # if i != j: # this line does not affect the outcome\n",
    "            count += 2\n",
    "            PCR_result, indexes = pcr_product(primers[i], primers[j], target_seq)\n",
    "            PCR_result_rotated, indexes_rotated = pcr_product(primers[i], primers[j], target_seq_rotated)\n",
    "            if indexes[2] != -1 and indexes_rotated[2] == indexes[2]:\n",
    "                print(f\"Forward primer {i} and reverse primer {j} produce seq of length {indexes[2]} in {target_name}\")\n",
    "            elif indexes[2] != -1:\n",
    "                print(f\"Forward primer {i} and reverse primer {j} produce seq of length {indexes[2]} in {target_name}\")\n",
    "            elif indexes_rotated[2] != -1:\n",
    "                print(f\"Forward primer {i} and reverse primer {j} produce seq of length {indexes_rotated[2]} in {target_name}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6ec056cd-9358-44e7-b192-fc00ca85b586",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward primer A and reverse primer D produce seq of length 2642 in HIV plasmid\n",
      "Forward primer A and reverse primer 5LTR produce seq of length 6833 in HIV plasmid\n",
      "test for symmetrical result\n",
      "Forward primer A and reverse primer D produce seq of length 2642 in HIV plasmid\n",
      "Forward primer A and reverse primer 5LTR produce seq of length 6833 in HIV plasmid\n"
     ]
    }
   ],
   "source": [
    "# define HIV specific primers, suggested\n",
    "# A and D are used like a control sample\n",
    "suggested_primers = ['A','D','LSS','LSP','3LTR','3LTRnest','5LTR','5LTRnest']\n",
    "target_seq = plasmid_HIV\n",
    "\n",
    "# run pcr\n",
    "run_pcr(suggested_primers, target_seq, 'HIV plasmid')\n",
    "print('test for symmetrical result')\n",
    "run_pcr(suggested_primers, target_seq[200:]+target_seq[:200], 'HIV plasmid')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7948ddd-cd00-4a38-be00-3fef4d52bc79",
   "metadata": {},
   "source": [
    "# PCR products for HIV primers on HXB2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "095480d0-0408-4485-a669-325af2d8b435",
   "metadata": {},
   "source": [
    "HXB2 is used as a reference for HIV genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0b66f616-712f-468d-8324-7c8628c36dc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read HXB2\n",
    "HXB2 = SeqIO.read('Sequences/HXB2_reference.fasta', 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7b700664-f08d-4026-9433-1c40eee7c21d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward primer A and reverse primer D produce seq of length 2642 in HXB2\n",
      "Forward primer A and reverse primer 5LTR produce seq of length 6836 in HXB2\n",
      "Forward primer 3LTR and reverse primer D produce seq of length 4397 in HXB2\n",
      "Forward primer 3LTR and reverse primer 5LTR produce seq of length 8591 in HXB2\n",
      "Forward primer 3LTRnest and reverse primer D produce seq of length 4380 in HXB2\n",
      "Forward primer 3LTRnest and reverse primer 5LTR produce seq of length 8574 in HXB2\n",
      "test for symmetrical result\n",
      "Forward primer A and reverse primer D produce seq of length 2642 in HXB2\n",
      "Forward primer A and reverse primer 5LTR produce seq of length 6836 in HXB2\n",
      "Forward primer 3LTR and reverse primer D produce seq of length 4397 in HXB2\n",
      "Forward primer 3LTR and reverse primer 5LTR produce seq of length 8591 in HXB2\n",
      "Forward primer 3LTRnest and reverse primer D produce seq of length 4380 in HXB2\n",
      "Forward primer 3LTRnest and reverse primer 5LTR produce seq of length 8574 in HXB2\n"
     ]
    }
   ],
   "source": [
    "# define HIV specific primers, suggested\n",
    "# A and D are used like a control sample\n",
    "# suggested_primers = ['A','D','LSS','LSP','3LTR','3LTRnest','5LTR','5LTRnest']\n",
    "suggested_primer = primers.keys()\n",
    "target_seq = str(HXB2.seq)\n",
    "\n",
    "# run pcr\n",
    "run_pcr(suggested_primers, target_seq, 'HXB2')\n",
    "print('test for symmetrical result')\n",
    "run_pcr(suggested_primers, target_seq[200:]+target_seq[:200], 'HXB2')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e0e46b4-73c2-479c-9f37-041d65bfbea8",
   "metadata": {},
   "source": [
    "# PCR products for HIV primers on HIV subtype A1 and A6"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c46d606-b763-4798-bf75-a0eadd87d2a8",
   "metadata": {},
   "source": [
    "In Kazakhstan and other FSU countries, HIV subtype A1 and A6 are the common occuring subtypes. The data with full genomes are available in Los Alamos National Laboratory (LANL) HIV Sequence Database https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html#subtype_6."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f9736e8c-449a-4279-ac18-7d83a32b5201",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ref.A1.AU.03.PS1044_Day0.DQ676872\n",
      "Ref.A1.RW.92.92RW008.AB253421\n",
      "Ref.A1.UG.92.92UG037_A40.AB253429\n",
      "Ref.A6.IT.02.60000.EU861977\n",
      "Ref.A6.UA.00.98UA0116.AF413987\n",
      "Ref.A6.UA.12.DEMA112UA024.KU749403\n"
     ]
    }
   ],
   "source": [
    "# print names of A1 and A6 subsequences\n",
    "hivs = SeqIO.parse('Sequences/HIV1_REF_2020_genome_DNA.fasta', 'fasta')\n",
    "\n",
    "for hiv in hivs:\n",
    "    if hiv.name.split('.')[1] in ['A1', 'A6']:\n",
    "        print(hiv.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "02d7be6f-241c-40b5-b9e8-9845066e8d94",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward primer A and reverse primer D produce seq of length 2642 in Ref.A1.RW.92.92RW008.AB253421\n",
      "Forward primer 3LTR and reverse primer D produce seq of length 4408 in Ref.A1.RW.92.92RW008.AB253421\n",
      "Forward primer A and reverse primer D produce seq of length 2642 in Ref.A1.UG.92.92UG037_A40.AB253429\n",
      "Forward primer A and reverse primer 5LTR produce seq of length 6878 in Ref.A1.UG.92.92UG037_A40.AB253429\n",
      "Forward primer 3LTR and reverse primer D produce seq of length 4413 in Ref.A1.UG.92.92UG037_A40.AB253429\n",
      "Forward primer 3LTR and reverse primer 5LTR produce seq of length 8649 in Ref.A1.UG.92.92UG037_A40.AB253429\n",
      "Forward primer A and reverse primer D produce seq of length 2642 in Ref.A6.IT.02.60000.EU861977\n",
      "Forward primer A and reverse primer D produce seq of length 2642 in Ref.A6.UA.00.98UA0116.AF413987\n",
      "Forward primer A and reverse primer D produce seq of length 2642 in Ref.A6.UA.12.DEMA112UA024.KU749403\n",
      "Forward primer 3LTR and reverse primer D produce seq of length 4411 in Ref.A6.UA.12.DEMA112UA024.KU749403\n"
     ]
    }
   ],
   "source": [
    "# define HIV specific primers, suggested\n",
    "# A and D are used like a control sample\n",
    "suggested_primers = ['A','D','LSS','LSP','3LTR','3LTRnest','5LTR','5LTRnest']\n",
    "\n",
    "# run PCR\n",
    "hivs = SeqIO.parse('Sequences/HIV1_REF_2020_genome_DNA.fasta', 'fasta')\n",
    "\n",
    "for hiv in hivs:\n",
    "    if hiv.name.split('.')[1] in ['A1', 'A6']:\n",
    "        target_seq = str(hiv.seq.replace(\"-\", \"\"))\n",
    "        \n",
    "        # run pcr\n",
    "        run_pcr(suggested_primers, target_seq, hiv.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a27906b3-24a6-4b13-83b1-a0cb2dd303c7",
   "metadata": {},
   "source": [
    "# Save primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a6aed015-e080-4ad6-a42f-c4bdde57f85d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>CAGGAGCAGATGATACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>CCCTTCACCTTTCCAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LSS</th>\n",
       "      <td>GTCCCTTAAGCGGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LSP</th>\n",
       "      <td>GTAATACGACTCACTATAGGGCCTCCGCTTAAGGGACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3LTR</th>\n",
       "      <td>TGTGACTCTGGTAACTAGAGATCCCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3LTRnest</th>\n",
       "      <td>GAGATCCCTCAGACCCTTTTAGTCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5LTR</th>\n",
       "      <td>TCAGGGAAGTAGCCTTGTGTGTGGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5LTRnest</th>\n",
       "      <td>CACTGTTGTCTTTTCTGGGAGTGAACTAGCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               0\n",
       "A                             CAGGAGCAGATGATACAG\n",
       "D                             CCCTTCACCTTTCCAGAG\n",
       "LSS                              GTCCCTTAAGCGGAG\n",
       "LSP       GTAATACGACTCACTATAGGGCCTCCGCTTAAGGGACT\n",
       "3LTR                 TGTGACTCTGGTAACTAGAGATCCCTC\n",
       "3LTRnest              GAGATCCCTCAGACCCTTTTAGTCAG\n",
       "5LTR                   TCAGGGAAGTAGCCTTGTGTGTGGT\n",
       "5LTRnest         CACTGTTGTCTTTTCTGGGAGTGAACTAGCC"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create dataframe\n",
    "df = pd.DataFrame([primers], index=[0])\n",
    "df.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f86f5787-57d6-4a04-8c44-53281ffe114c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to csv\n",
    "df.to_csv('Sequences/primers.csv', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "aaa5fb37-9624-4270-a7a8-dcd8f26590e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1280"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b42825f-af9e-4941-9a75-983970414c4d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
